Redesign field properties to work with nested structure

ongchi opened the pull request:
Fixes #7

ongchi committed:
make deserialize_any return error

fixed width is not self-describing format should avoid relying on deserialize_any and deserialize_ignored_any.

ongchi committed:
update test cases

Deserialize would infer field name from type definition except for HashMap, remove name specify on fields to make test cases more proper.

ongchi committed:
Implement FieldSet and add some test cases

ongchi committed:
Update reader.rs

ongchi committed:
Refactor FieldSet

Refactor FieldSet and make everything works.

ongchi committed:
Create macros.rs

ongchi committed:
cleanup

twking7 commented on fixed_width/src/de.rs:
```diff
@@ -488,29 +485,17 @@ impl<'a, 'de: 'a> serde::Deserializer<'de> for &'a mut Deserializer<'de> {
-        }
+    // FixedWidth is not self describing format should avoid this method.
+    fn deserialize_any<V: Visitor<'de>>(self, _visitor: V) -> Result<V::Value, Self::Error> {
+        Err(DeserializeError::WontImplement)
```
I agree this method probably doesn't make a ton of sense for a non self describing format, what do you think about deserializing to a string if the user asks for an Any?

ongchi replied:
There are no ways to infer the type at runtime as I know.
I think most people use derive macro to implement `Deserialize` and `Visitor` trait automatically, and the `Visitor` will forward `Deserializer` to a specific method by the type signature. Also the same story when implementing required traits manually. The format specifications for fixed-width should determine at compile-time rather than at runtime, and it's the only way to go. If the target type signature is `String`, just pass the `Deserializer` to `deserialize_string` but not `deserialize_any`. For more specific usage, people probably should go with field attribute (`#[serde(deserialize_with = "path")]`) to specify their `Deserializer` for that case.

twking7 replied:
Fair enough. FWIW, when I originally wrote this I took inspiration from the serde implementation from csv (which ultimately calls this function): https://github.com/BurntSushi/rust-csv/blob/master/src/deserializer.rs#L206

ongchi replied:
Took some time to figure out rust-csv, I could be wrong, but just leave a note.
The `deserialize_any` is called by `EnumAccess` to determine its variant in `rust-cvs`.
The cvs format is created by hands or machine generated, may contains unexpected values in a column in my experience, so I can understand why the value would pass to `deserialize_any`, but implement global enum maybe a better choice (like [serde_json::value::Value](https://docs.serde.rs/serde_json/value/enum.Value.html)), but also, may be too heavy for that case.
For `fixed_width`, my use case is parsing data generated from fortran fixed format output, its machine generated contents, guarantee typed, and everything should known at compile-time.

twking7 commented on fixed_width/src/de.rs:
```diff
@@ -521,10 +506,13 @@ impl<'a, 'de: 'a> de::SeqAccess<'de> for &'a mut Deserializer<'de> {
+                let mut de = Deserializer::new(self.input, self.fields.next().unwrap());
+                seed.deserialize(&mut de).map(Some)
+            }
+            None => Ok(None),
```
Nice!

ongchi commented:
I just realized it's a bit messy to put complex syntax into procedural macro and might not be a good choice. The current macro is nice and clear for simple cases, adding support for nested fields definition seems too verbose for lengthy fields, maybe I should just implement a container attribute letting user specify their complex fields by a custom function. ex:

```rust
#[derive(FixedWidth, Deserialize)]
#[fixed_width(field_def="custom_function")]
struct Foo {
    ...
}
```

Which is identically with implement the `FixedWidth` trait manually but more like serde api, or just leave it untouched.
What do you think?

ongchi committed:
Update macros.rs

ongchi committed:
Update field definitions

- rename Field to FieldConfig
- add append method to FieldSet
- add extend method to FieldSet

ongchi committed:
Update derive macro

twking7 commented:
@ongchi I think that makes sense!

ongchi committed:
Update documents

ongchi commented:
> I just realized it's a bit messy to put complex syntax into procedural macro and might not be a good choice. The current macro is nice and clear for simple cases, adding support for nested fields definition seems too verbose for lengthy fields, maybe I should just implement a container attribute letting user specify their complex fields by a custom function. ex:
> 
> ```rust
> #[derive(FixedWidth, Deserialize)]
> #[fixed_width(field_def="custom_function")]
> struct Foo {
>     ...
> }
> ```
> 
> Which is identically with implement the `FixedWidth` trait manually but more like serde api, or just leave it untouched. What do you think?

Maybe someday will get this done

ongchi commented:
Document updated. Everything should works properly.

twking7 commented:
Thank you! I will check this out asap - been busy with some IRL stuff.

twking7 commented on README.md:
```diff
@@ -1,17 +1,19 @@
+Fixed width in this case means that each line of the file is the same number of bytes.
+It supports unicode, but only if the unicode fits in the designated byte size of the record.
+It also provides a few useful abstractions to ease serializing and deserializing data into and out
+of fixed width files.
```
👍

twking7 reviewed (approved):
Thank you for contributing this! It looks great. I'll merge this and push the new version.
